### PR TITLE
Add debug switch for hydro allocation

### LINKED ISSUE #253: Give the possibility to extract hydro pre-allocation results in debug mode
**Need**
Today, the results of the pre-allocation hydro heuristics are not displayed to the used. However, they are already calculated during Antares solving and they could be easily displayed in additional results files (the generation of these files has already been developed in Antares for past debug purposes). It would be useful for result analysis to allow the Antares R&D team to activate the generation of these files if needed.

**Proposed enhancement**
Add a new "hydro-debug" boolean parameter in the "generaldata.ini" file in order to activate or deactivate the generation of these hydro pre-allocation result files when running an Antares simulation.

### PR DESCRIPTION
**Pull request by flomnes**

### CODE REVIEW

**Comment by flomnes** [comment]
I added a switch called "hydro-debug" in the general section of file **generaldata.ini**. It allows the user to enable/disable an existing debug output.

There are some issues with the output files :
- In `daily.<area>.txt`, when `reservoirManagement = true`, the first day of each month has 6 more columns
- The format for `daily.<area>.txt` changes based on the value of `reservoirManagement`

When `reservoirManagement = true`, `daily.<area>.txt` looks like
```
	Niveau init : 0.100000

-------------
January
-------------
			NiveauMin	Apports		TurbMax		TurbCible	TurbCible MAJ	Niveaux D	Niveaux F	Turbines	Overflows	Deviations	Violations	Deviation Max	Violation Max	Waste		Cout total	Turb mois no previous W		Turb mois + previous W 
0		1	10.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000	0.000000		0.000000		0.000000	-0.096875	0.000000					0.000000
1		2	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
2		3	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
3		4	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
4		5	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
5		6	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
6		7	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
7		8	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
8		9	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
9		10	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
10		11	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
11		12	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
12		13	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
13		14	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
14		15	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
15		16	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
16		17	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
17		18	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
18		19	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
19		20	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
20		21	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
21		22	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
22		23	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
23		24	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
24		25	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
25		26	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
26		27	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
27		28	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
28		29	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
29		30	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000
30		31	0.000000	0.000000	1200.000000	0.000000	0.000000		10.000000	10.000000	0.000000	0.000000	0.000000	0.000000

-------------
February
-------------
...
```
When `reservoirManagement = false`, `daily.<area>.txt` looks like
```
	Turbine			OPP				Turbine Cible	DLE				DLN
0	3601.062141	24000.000000	3601.062141	97856.000100	75229.000000
1	793.019244	24000.000000	793.019244	2227.000100	-20400.000000
2	2973.263556	24000.000000	2973.263556	60617.000100	37990.000000
3	3785.072817	24000.000000	3785.072817	110840.000100	88213.000000
4	3802.863141	24000.000000	3802.863141	112147.000100	89520.000000
5	0.912422	24000.000000	0.912422	0.000100	-22627.000000
6	2807.622727	24000.000000	2807.622727	52524.000100	29897.000000
7	4484.137651	24000.000000	4484.137651	169320.000100	146693.000000
8	4367.275875	24000.000000	4367.275875	158503.000100	135876.000000
9	4589.683069	24000.000000	4589.683069	179460.000100	156833.000000
10	4189.102374	24000.000000	4189.102374	142828.000100	120201.000000
11	3793.908756	24000.000000	3793.908756	111488.000100	88861.000000
12	4448.833911	24000.000000	4448.833911	166007.000100	143380.000000
13	4267.676506	24000.000000	4267.676506	149620.000100	126993.000000
14	4406.522351	24000.000000	4406.522351	162088.000100	139461.000000
15	4173.891656	24000.000000	4173.891656	141535.000100	118908.000000
...
```
@PPlessiez can you tell us what values are required in each case ?

**Comment by PPlessiez** [comment]
The main differences between the hydro behaviour when reservoir mode is activated compared to when it's not are the following:
- With RM, monthly hydro energy targets are determined from the yearly inflows, then daily energy targets are determined from these monthly energy targets. Without RM, monthly inflows are directly considered as the monthly energy target, so the monthly hydro allocation is skipped.
- With RM, the reservoir size and the rule curves are properly taken into account. Without RM, the reservoir size is considered infinite and the rule curves are ignored.
These two considerations might explain the main differences between the two files.

So with RM, my best guess regarding the columns signification would be the following:
- Niveau min: minimum reservoir level as set by the user (lower rule curve)
- Apports: daily inflows
- TurbMax: maximum generating energy over the whole day as set by the user
- TurbCible: daily energy target as resulting from splitting the monthly energy target using the residual load on each day
- TurbCible MAJ : daily energy target updated from TurbCible, taking into account reservoir constraints
- Niveaux D: reservoir level at the beginning of the day
- Niveaux F: reservoir level at the end of the day (= Niveaux D + Apports - TurbCible MAJ)
- Turbines: ?
- Overflows: energy overflowing from the reservoir on that day (if reservoir level goes bayond 100%)
- Deviations: deviation of TurbCible MAJ compared to TurbCible
- Violations: violations of the rule curves
- Deviation max: max value of deviation over the whole month
- Violation max: max value of violation over the whole month
- Waste: ?
- Coût total: cost of generating energy on that day (if water values are provided)?
- Turb mois no previous W: energy generated over the whole month?
- Turb mois + previous W: energy generating over the whole month taking into account the previous week?

Without RM:
Here, this is way less clear. I would say that "Cible" is equivalent to "TurbCible" from the list above, but I have no idea what the other variables refer to.

To proceed with this change, I would suggest that this version is released so @JMJ-rte and I can test it and understand what these variables mean. In the meantime, @flomnes and @guilpier can also investigate the meaning of these parameters from the code.

**Comment by flomnes** [comment]
@PPlessiez @JMJ-rte 

The compiled package can be found [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/suites/2827263544/artifacts/63013419).

Enable debug in your study by adding 
```ini
[output]
hydro-debug = true
```
in **generaldata.ini**.

The debug files will be placed in **output\<simulation>\debug\solver**.

File: src/solver/hydro/management/daily.cpp [hunk 1]
```diff
@@ -58,6 +57,170 @@ enum
     maxDTG = 32
 };
 
+struct DebugData
+{
+    using PerArea = HydroManagement::PerArea;
+    using InflowsT = Matrix<double, Yuni::sint32>::ColumnType;
```
Comment (guilpier-code) [inline src/solver/hydro/management/daily.cpp]: I would not hesitate to rename **InflowsT** into **InflowsType**.
Same remark for the other types.

File: src/solver/hydro/management/daily.cpp [hunk 2]
```diff
@@ -112,9 +275,8 @@ inline void HydroManagement::prepareDailyOptimalGenerations(Solver::Variable::St
             assert(dYear < 366);
             scratchpad.optimalMaxPower[dYear] = maxP[dYear];
 
-#if HYDRO_DAILY_SOLVER_DEBUG != 0
-            debugOPP[dYear] = maxP[dYear] * maxE[dYear];
-#endif
+            if (debugData != nullptr)
```
Comment (guilpier-code) [inline src/solver/hydro/management/daily.cpp]: `if (debugData)` is enough.
Same remark for other occurrences of ` if (debugData != nullptr)`

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/hydro/management/daily.cpp (hunk 2 above)
Comment (flomnes) [reply src/solver/hydro/management/daily.cpp]: done

File: src/solver/hydro/management/daily.cpp (hunk 1 above)
Comment (flomnes) [reply src/solver/hydro/management/daily.cpp]: done

**Review by guilpier-code — APPROVED** [review]

**Pull request merged**